aten/src/ATen/core/LegacyTypeDispatch.h

```diff
@@ -140,18 +140,45 @@ class CAFFE2_API LegacyTypeDispatch {
 
 CAFFE2_API LegacyTypeDispatch& globalLegacyTypeDispatch();
 
+struct CAFFE2_API NonVariableTypeMode {
+  static bool is_enabled();
+  static void set_enabled(bool enabled);
+};
+
+// A RAII, thread local (!) guard that has the following effect:
+//
+// Upon construction: sets NonVariableTypeMode_enabled for the current thread to
+// control whether we are in non-Variable-type mode.
+//
+// Upon destruction: sets NonVariableTypeMode_enabled back to the original value.
+//
+// See NOTE [ Treating Variables as non-Variables in type dispatch ] for details.
+struct CAFFE2_API AutoNonVariableTypeMode {
+  AutoNonVariableTypeMode(bool enabled) : prev_mode(NonVariableTypeMode::is_enabled()) {
+    NonVariableTypeMode::set_enabled(enabled);
+  }
+  ~AutoNonVariableTypeMode() {
+    NonVariableTypeMode::set_enabled(prev_mode);
+  }
+  bool prev_mode;
+};
+
 /**
  * Return the Type object corresponding to this Tensor, which we can
  * use to do dynamic dispatch to operators from.  This method is NOT
  * intended to be used by end-users; it is purely an implementation
  * detail.
+ *
+ * NOTE: We also check `at::NonVariableTypeMode`, and if it's enabled
+ * we always return non-Variable type in this function.
+ * See NOTE [ Treating Variables as non-Variables in type dispatch ]
  */
 inline Type& legacyTensorType(const TensorImpl& tensor) {
   // NB: It's valid to use getTypeRaw here, because the TensorImpl
   // could not have been created without initializing the Type first.
   // TODO: This is not actually true via the Caffe2 codepath!  Make
   // it so.
-  return *globalLegacyTypeDispatch().getTypeRaw(tensorTypeIdToBackend(tensor.type_id()), typeMetaToScalarType(tensor.dtype()), tensor.is_variable());
+  return *globalLegacyTypeDispatch().getTypeRaw(tensorTypeIdToBackend(tensor.type_id()), typeMetaToScalarType(tensor.dtype()), tensor.is_variable() && !at::NonVariableTypeMode::is_enabled());
```

Comment:
is this actually doing anything different than globalContext().getType(tensor)?

Reply:
I don't think it's doing anything different than `globalContext().getType(tensor)`. We might be able to get rid of `legacyTensorType(tensor)` and just use `globalContext().getType(tensor)` in the future.